cl_manycore/libraries/bsg_manycore_cuda.h

```diff
@@ -15,27 +15,289 @@
+ * @param[in]  device        Pointer to device
+ * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
+ */
+__attribute__((warn_unused_result))
+int hb_mc_device_tile_groups_execute (device_t *device);
+
+
+
+/**
+ * Checks to see if all tile groups in a device are finished.
+ * @param[in]  device        Pointer to device
+ * returns HB_MC_SUCCESS if all tile groups are finished, and HB_MC_FAIL otherwise.
+ */
+__attribute__((warn_unused_result))
+int hb_mc_device_all_tile_groups_finished(device_t *device);
+
+
+
+/**
+ * Waits for a tile group to send a finish packet to device.
+ * @param[in]  device        Pointer to device
+ * return HB_MC_SUCCESS after a tile group is finished, gets stuck in infinite look if no tile group finishes.
+ */
+__attribute__((warn_unused_result))
+int hb_mc_device_wait_for_tile_group_finish_any(device_t *device);
+
+
+
+
+/**
+ * Copies a buffer from src on the host/device DRAM to dst on device DRAM/host.
+ * @param[in]  device        Pointer to device
+ * @parma[in]  src           EVA address of src 
+ * @param[in]  name          EVA address of dst
+ * @param[in]  count         Size of buffer to be copied
+ * @param[in]  hb_mc_memcpy_kind         Direction of copy (hb_mc_memcpy_to_device / hb_mc_memcpy_to_host)
+ * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
+ */
+__attribute__((warn_unused_result))
+int hb_mc_device_memcpy (device_t *device, void *dst, const void *src, uint32_t count, enum hb_mc_memcpy_kind kind);
+
+
+
+
+
+
+/**
+ * Takes in a device_t struct and initializes a mesh of tile in the Manycore device.
+ * @param[in]  device        Pointer to device
+ * @parma[in]  dim           X/Y dimensions of the tile pool (mesh) to be initialized
+ * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
+ */
+__attribute__((warn_unused_result))
+int hb_mc_mesh_init (device_t *device, hb_mc_dimension_t dim); 
+
+
+
+
+/**
+ * Takes the grid size, tile group dimensions, kernel name, argc, argv* and the finish signal address, calls hb_mc_tile_group_enqueue to initialize all tile groups for grid.
+ * @param[in]  device        Pointer to device
+ * @param[in]  grid_dim      X/Y dimensions of the grid to be initialized
+ * @param[in]  tg_dim        X/Y dimensions of tile groups in grid
+ * @param[in]  name          Kernel name to be executed on tile groups in grid
+ * @param[in]  argc          Number of input arguments to kernel
+ * @param[in]  argv          List of input arguments to kernel
+ * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
+ */
+__attribute__((warn_unused_result))
+int hb_mc_grid_init (device_t *device, hb_mc_dimension_t grid_dim, hb_mc_dimension_t tg_dim, char *name, uint32_t argc, uint32_t argv[]);
+
+
+
+
+/**
+ * Searches for a free tile group inside the device mesh and allocoates it, and sets the dimensions, origin, and id of tile group.
+ * @param[in]  device        Pointer to device
+ * @param[in]  tg            Pointer to tile group
```

Comment:
Is `tg` an input or an output?

Reply:
It is an input. The function tries to find a group of free tiles for this specific tile group, and once it's found it sends packets to set the tile symbols using the data inside tg struct.